src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,343 @@
+  constraint.Impostors(impostors);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+};
+
+//! Calculate cost over batches.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
+                                          const size_t begin,
+                                          const size_t batchSize)
+{
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+}
+
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = 0; i < dataset.n_rows; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
```

Comment:
I think that you could also move the `gradient += (1 - regularization) * OuterProduct(cij)` outside the loop, and just take the sum of all `cij`s inside the loop.